docs(core): edit and add docs for ng-container (#44762)
angular/angular: aio/content/special-elements/core/ng-container.md

Angular's `<ng-container>` element is a special element which doesn't actually gets added to the DOM but can simply be used to hold structural directives.

> The `<ng-container>` is very useful for utilizing structural directives without adding new unnecessary elements to the DOM, this not only increases performance (even so slightly) since the browser ends up rendering less elements but can also be a valuable asset in having cleaner DOMs and styles alike.

Author: yeah, I know what you mean, but this is how these two paragraphs are rendered:
![Screenshot at 2022-01-20 20-19-06](https://user-images.githubusercontent.com/61631103/150415840-7136332e-e8c6-4bb9-99b2-a9d18c1d2dbc.png)

the first one is the quick summary text for the element, and the second one is sort of the same but goes a bit more in detail.

:thinking: 

I'll try to keep the structure as is (quick summary + a bit more details) but try to make things less redundant :slightly_smiling_face:

> The `<ng-container>` is very useful for utilizing structural directives without adding new unnecessary elements to the DOM, this not only increases performance (even so slightly) since the browser ends up rendering less elements but can also be a valuable asset in having cleaner DOMs and styles alike.

Author: I've tried to improve things a bit, but it wasn't too straightforward given that I wanted to keep this structure, so I made the initial text very strict and to the point and change a bit the description text to make it a bit more different from the initial one

I hope this is an improvement, please let me know what you think :slightly_smiling_face:

@usageNotes

### With `*NgIf`s

One common use case of `<ng-container>` is alongside the `*ngIf` structural directive, by using the special element we can produce very clean templates easy to understand and work with.

For example we may want to have a number of elements shown conditionally but they do not need to be all under the same root element, that can be easily done by wrapping them in such a block:
```html
  <ng-container *ngIf="condition">
    ...
  </ng-container>
```

This can also be augmented with the an else statement alongside an `<ng-template>` as:
```html
  <ng-container *ngIf="condition; else templateA">
    ...
  </ng-container>
  <ng-template #templateA>
    ...
  </ng-template>
```

> Or we can even only use the `<ng-container>` to redirect to the correct views without even needing to prove any content for it:

Author: sure I'll remove it :slightly_smiling_face::+1:

```html
  <ng-container *ngIf="condition; then templateA else templateB"></ng-container>
  <ng-template #templateA>
    ...
  </ng-template>
  <ng-template #templateB>
    ...
  </ng-template>
```

> ### Cleaner DOMs and Styles

Author: I wanted to drive home the point of `ng-container`s helping with styles, if you think I am overdoing it here I guess I'll try to summarise the point in a sentence somewhere :slightly_smiling_face:

> ### Cleaner DOMs and Styles

Author: done, I added a sentence about styling in the description above, I hope it does not look too much out of place, please let me know what you think :slightly_smiling_face:

Let's for example take the following HTML:

```html
<div class="flex-container">
  <ng-container *ngIf="numbers">
    <span>1</span>
    <span>2</span>
    <span>3</span>
  </ng-container>
  <ng-container *ngIf="letters">
    <span>A</span>
    <span>B</span>
    <span>C</span>
  </ng-container>
</div>
```

Which gets styled by the following css code:
```css
.flex-container {
  display: flex;
  justify-content: space-between;
}
```

Note how valuable is here not to have extra elements added to the DOM, thanks to that our flex container can apply its `justify-content` to the correct elements, no matter on whether both `*ngIf`'s conditions satisfied, just one or none, if we would have had to use `div`s instead of `ng-container` this would have been much more difficult since we would also need to make such `div`s flex containers as well and make them match the parent's layout accordingly.

Naturally this not only applied to flex containers, but to any other css styling which depends on the actual DOM structure (as for example `grid`, margins, the child combinator selector, etc...).

### Combination of multiple structural directives

> Multiple structural directives cannot be used on the same element, so if you need to take advantage of more than one structural directive it is advised to use an `<ng-container>` per structural directive.

Author: yeah, I've tried looking around but I couldn't really find anything, the only candidate could be:
https://angular.io/guide/built-in-directives#one-per-element , but it already mentions `<ng-container>` (which auto-links to this api docs page)

(the [structural directives guide](https://angular.io/guide/structural-directives) does not seem to have anything about this)

what do you think? maybe it could be good to add a section about this in the guide? if that sounds like a good idea I can do it here or in a separate PR, whatever works best for you :slightly_smiling_face:

> Multiple structural directives cannot be used on the same element, so if you need to take advantage of more than one structural directive it is advised to use an `<ng-container>` per structural directive.

Author: Sure you can leave it with me! :muscle::smile: 

___

By the way (a bit off topic), I don't think we have a "structural directive" page per se, the current one: https://angular.io/guide/structural-directives

Is actually titled "Writing structural directives", if possible in a future PR I would love to change it up a bit so that it could actually be a page generally dedicated to structural directives and the writing of them could be a part of it.

Note that in the current page there are generic sections such as [Structural Directive Shorthand](https://angular.io/guide/structural-directives#structural-directive-shorthand) and [Structural Directive Syntax](https://angular.io/guide/structural-directives#structural-directive-syntax-reference), but they (at least to me) actually feel out of place given the page's title and main focus.

> Multiple structural directives cannot be used on the same element, so if you need to take advantage of more than one structural directive it is advised to use an `<ng-container>` per structural directive.

Reviewer: @dario-piotrowicz that's a great idea. I was a bit confused yesterday while looking for a "generic" structural directives page. I like you idea, that makes a lot of sense. Could you please create a ticket to describe that (so we don't lose it)?

> Multiple structural directives cannot be used on the same element, so if you need to take advantage of more than one structural directive it is advised to use an `<ng-container>` per structural directive.

Author: of course! :slightly_smiling_face:

The most common scenario is with `*ngIf` and `*ngFor`, lets for example imagine that we have a list of items but each item needs to be showed only if a certain condition is true, we could be tempted to try something like:
```html
<ul>
  <li *ngFor="let item of items" *ngIf="item.isValid">
    {{ item.name }}
  </li>
</ul>
```

As we said that would not work, so what we can do is to simply move one of the structural directives to an `<ng-container>` element which would then wrap the other one, like so:
```html
<ul>
  <ng-container *ngFor="let item of items">
    <li *ngIf="item.isValid">
      {{ item.name }}
    </li>
  </ng-container>
</ul>
```

This would work as intended without introducing any new unnecessary elements in the DOM.

### Use alongside ngTemplateOutlet

`NgTemplateOutlet` directives can be applied to any element but they are most of the time applied
to `<ng-container>` ones since by combining the two we get a very clear and easy to follow HTML and DOM
structure in which no extra elements are necessary present and template views are instantiated where requested.

> You can see an example of this in the [`NgTemplateOutlet`s api documentation page](/api/common/NgTemplateOutlet).

Reviewer: I think it'd be great to have a quick example here, but also keep a reference to the `NgTemplateOutlet` API page to learn more.

Author: I've added an example, please have a look it is ok :smiley:

A special element that can hold structural directives without adding new elements to the DOM.

The `<ng-container>` allows us to use structural directives without any extra element, making sure that the only DOM changes being applied are those dictated by the directives themselves.

This not only increases performance (even so slightly) since the browser ends up rendering less elements but can also be a valuable asset in having cleaner DOMs and styles alike.

It can for example enable us to use structural directives without breaking styling dependent on a precise DOM structure (as for example the ones we get when using flex containers, margins, the child combinator selector, etc...).

@usageNotes

### With `*NgIf`s

> One common use case of `<ng-container>` is alongside the `*ngIf` structural directive, by using the special element we can produce very clean templates easy to understand and work with.

Reviewer 2: structural directive, by using the special element
```suggestion
Break up this sentence into two: "One common use case of `<ng-container>` is alongside the `*ngIf` structural directive. By using the special element we can produce very clean templates easy to understand and work with."
```

> For example we may want to have a number of elements shown conditionally but they do not need to be all under the same root element, that can be easily done by wrapping them in such a block:

Reviewer 2: For example we may want to have a number of elements shown conditionally but they do not need to be all under the same root element, that can be easily done by wrapping them in such a block:
```suggestion
1. Insert comma after "For example":  "For example, we may want to have a number of elements shown conditionally but they do not need to be all under the same root element, that can be easily done by wrapping them in such a block:

2. Break this sentence up into two separate ones: "For example, we may want to have a number of elements shown conditionally but they do not need to be all under the same root element. That can be easily done by wrapping them in such a block:
```

```html
  <ng-container *ngIf="condition">
    ...
  </ng-container>
```

This can also be augmented with the an else statement alongside an `<ng-template>` as:
```html
  <ng-container *ngIf="condition; else templateA">
    ...
  </ng-container>
  <ng-template #templateA>
    ...
  </ng-template>
```

### Combination of multiple structural directives

> Multiple structural directives cannot be used on the same element, so if you need to take advantage of more than one structural directive it is advised to use an `<ng-container>` per structural directive.

Reviewer 2: Multiple structural directives cannot be used on the same element, so if you need to take advantage of more than one structural directive it is advised to use an `<ng-container>` per structural directive.
```suggestion

1. Replace the comma after "same element" with a semicolon and add a comma after "structural directive" : "Multiple structural directives cannot be used on the same element; so if you need to take advantage of more than one structural directive, it is advised to use an `<ng-container>` per structural directive.

2. Delete "so":   "Multiple structural directives cannot be used on the same element; if you need to take advantage of more than one structural directive, it is advised to use an `<ng-container>` per structural directive.
```

> The most common scenario is with `*ngIf` and `*ngFor`, for example let imagine that we have a list of items but each item needs to be displayed only if a certain condition is true. We could be tempted to try something like:

Reviewer 2: The most common scenario is with `*ngIf` and `*ngFor`, for example let imagine that we have a list of items but each item needs to be displayed only if a certain condition is true. We could be tempted to try something like:
```suggestion
1. Break this up into two sentences 
2.  Add a comma after "For example"
3. Change "let imagine" to let's imagine" 

"The most common scenario is with `*ngIf` and `*ngFor`. For example, let's imagine that we have a list of items but each item needs to be displayed only if a certain condition is true. We could be tempted to try something like:
```

Author: silly typo! :man_facepalming:

```html
<ul>
  <li *ngFor="let item of items" *ngIf="item.isValid">
    {{ item.name }}
  </li>
</ul>
```

> As we said that would not work, so what we can do is to simply move one of the structural directives to an `<ng-container>` element which would then wrap the other one, like so:

Reviewer 2: As we said that would not work, so what we can do is to simply move one of the structural directives to an `<ng-container>` element which would then wrap the other one, like so:
```suggestion
Insert comma after "element" and delete "so", for clarity: "As we said that would not work,  what we can do is to simply move one of the structural directives to an `<ng-container>` element , which would then wrap the other one, like so:
```

```html
<ul>
  <ng-container *ngFor="let item of items">
    <li *ngIf="item.isValid">
      {{ item.name }}
    </li>
  </ng-container>
</ul>
```

This would work as intended without introducing any new unnecessary elements in the DOM.

### Use alongside ngTemplateOutlet

> The `NgTemplateOutlet` directive can be applied to any element but most of the time it's applied

Reviewer 2: The `NgTemplateOutlet` directive can be applied to any element but most of the time it's applied
to `<ng-container>` ones since by combining the two we get a very clear and easy to follow HTML and DOM structure in which no extra elements are necessary present and template views are instantiated where requested.
```suggestion
1. Break up into two sentences
2. Insert comma after "combining the two"
3. Change "necessary" to "necessarily"

The `NgTemplateOutlet` directive can be applied to any element but most of the time it's applied
to `<ng-container>` ones.  By combining the two, we get a very clear and easy to follow HTML and DOM structure in which no extra elements are necessarily present and template views are instantiated where requested.
```

Author: The `necessary present` part was a left over from other changes, this suggestion made me notice that 
:sweat: :slightly_smiling_face::+1: 

`present` should not be there, so I just removed that, leaving us with:
`... no extra elements are necessary and ...`

@TMDavisGoogle I hope this is fine :slightly_smiling_face:

to `<ng-container>` ones since by combining the two we get a very clear and easy to follow HTML and DOM structure in which no extra elements are necessary present and template views are instantiated where requested.

> Let's for example imagine a situation in which we have a large HTML in which a small portion needs to be repeated in different places. A simple solution is to define an `<ng-template>` containing our repeating HTML and render that where necessary by using `<ng-container>` alongside an `NgTemplateOutlet`.

Reviewer 2: Let's for example imagine a situation in which we have a large HTML in which a small portion needs to be repeated in different places. A simple solution is to define an `<ng-template>` containing our repeating HTML and render that where necessary by using `<ng-container>` alongside an `NgTemplateOutlet`.
```suggestion
Switch word order at beginning of the sentence and insert comma after "For example" and "large HTML", for clarity:

For example, imagine a situation in which we have a large HTML, in which a small portion needs to be repeated in different places. A simple solution is to define an `<ng-template>` containing our repeating HTML and render that where necessary by using `<ng-container>` alongside an `NgTemplateOutlet`.
```

```html
<ul>
  <ng-container *ngFor="let item of items">
    <li *ngIf="item.isValid">
      {{ item.name }}
    </li>
  </ng-container>
</ul>
```

This would work as intended without introducing any new unnecessary elements in the DOM.

### Use alongside ngTemplateOutlet

The `NgTemplateOutlet` directive can be applied to any element but most of the time it's applied
to `<ng-container>` ones since by combining the two we get a very clear and easy to follow HTML and DOM structure in which no extra elements are necessary present and template views are instantiated where requested.

Let's for example imagine a situation in which we have a large HTML in which a small portion needs to be repeated in different places. A simple solution is to define an `<ng-template>` containing our repeating HTML and render that where necessary by using `<ng-container>` alongside an `NgTemplateOutlet`.

Like so:

```html
<!-- ... -->

<ng-container [ngTemplateOutlet]="importantTextTemplate"
              [ngTemplateOutletContext]="{ $implicit: relevantInfo }">

> </ng-container>

Author: Ah yeah :smiley: 

Sorry I wasn't actually aware of this shorthand syntax :sweat_smile: 

I've applied the change, but I have put the closing tag on a new line, so that it can help a bit not to get the line wrapped (not a big difference since the examples width shrinks and the line may need to be wrapped anyways, but still I think it helps)

I hope it's fine :slightly_smiling_face:

> </ng-container>

Reviewer: Yeah, looks great, thanks 👍

<!-- ... -->

> <ng-container [ngTemplateOutlet]="importantTextTemplate"
>               [ngTemplateOutletContext]="{ $implicit: urgentAnnouncement }">
> </ng-container>

Reviewer: Just to complete the first case above:
```suggestion
<ng-container *ngTemplateOutlet="tmpl; context: {$implicit: 'World'}"></ng-container>
```
Note: we can probably come up with a better values (instead of "Hello" and "World") :) I just wanted to simplify an example as much as possible.

Author: done :slightly_smiling_face: 

About the "Hello" and "World" values I think they are fine since we are keeping this so basic/simple, there's no much benefit in trying to make it more realistic/adding any extra meaning to things I guess :slightly_smiling_face:
